### PR TITLE
feat: detect iframe error for in-app heatmaps

### PR DESCRIPTION
**Pull request by pauldambra**
Not all sites can be embedded in an iframe... 

* tell people when their site is slow to load
* tell people when their site fails to load

only use timeouts and messages from the toolbar to detect errors

### CODE REVIEW

File: frontend/src/scenes/heatmaps/heatmapsBrowserLogic.ts [hunk 1]
```diff
@@ -50,6 +50,7 @@ export const heatmapsBrowserLogic = kea<heatmapsBrowserLogicType>([
         // TRICKY: duplication ends
         setIframeWidth: (width: number | null) => ({ width }),
         toggleFilterPanelCollapsed: true,
+        setIframeError: (error: string | null) => ({ error }),
```
Comment (pauldambra) [inline frontend/src/scenes/heatmaps/heatmapsBrowserLogic.ts:53]: this is a string since i'm going to follow-up with a change that adds an error if we time-out trying to load the iframe

File: frontend/src/scenes/heatmaps/heatmapsBrowserLogic.ts [hunk 2]
```diff
@@ -242,6 +249,27 @@ export const heatmapsBrowserLogic = kea<heatmapsBrowserLogicType>([
...
             // TODO: Add a timeout - if we haven't received a message from the iframe in X seconds, show an error
+
+            // Attempt to access iframe content
+            // this lets us probe for errors like X-Frame-Origin issues
+            try {
+                const theIframe = props.iframeRef.current
+                const iframeContent = theIframe?.contentWindow?.document || theIframe?.contentDocument
+
+                // Check if we can access iframe content
+                if (iframeContent?.body.innerHTML) {
```
Comment (benjackwhite) [inline frontend/src/scenes/heatmaps/heatmapsBrowserLogic.ts:260]: I don't think this is what we want 🤔 Given that we will be posting messages back and forth, we don't actually want to ever be loading the iframe in such a way that we can access the raw html do we?

Comment (benjackwhite) [reply frontend/src/scenes/heatmaps/heatmapsBrowserLogic.ts:260]: I was imagining we would do this based on a timeout of not getting the "ready" state from the toolbar after N seconds

Comment (pauldambra) [reply frontend/src/scenes/heatmaps/heatmapsBrowserLogic.ts:260]: 🤔 this is just an attempt to probe that the iframe actually loaded 

I _thought_ that we'd always be able to access the raw html from JS but I've been wrong before 🙈 

Do you already have an alternative? I didn't see a signal from the iframe that i could hook into

Comment (pauldambra) [reply frontend/src/scenes/heatmaps/heatmapsBrowserLogic.ts:260]: was trying to avoid having to wait for onLoad to timeout when the user can already see the request has failed

File: frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx [hunk 3]
```diff
@@ -211,26 +234,28 @@ function EmbeddedHeatmapBrowser({
...
-                // this is a very loose sandbox,
-                // but we specify it so that at least other capabilities are denied
-                sandbox="allow-scripts allow-same-origin"
-                // we don't allow things such as camera access though
-                allow=""
-            />
-
-            {loading && <SpinnerOverlay />}
+            <div className="relative flex-1 w-full h-full">
+                {iframeBanner && <IframeErrorOverlay />}
```
Comment (daibhin) [inline frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx:238]: Feels easier to pass this value through rather than calling the logic again in the component:
```suggestion
                {iframeBanner && <IframeErrorOverlay iframeBanner={iframeBanner} />}
```

File: frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx [hunk 4]
```diff
@@ -211,26 +234,28 @@ function EmbeddedHeatmapBrowser({
...
+                <LoadingOverlay />
```
Comment (daibhin) [inline frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx:239]: It's a personal rule but I think it's always better to conditionally render a component outside the component:
```suggestion
                {loading && <LoadingOverlay />}
```

**Review by daibhin — APPROVED** [review]
Couple of style things but nothing major

File: frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx (hunk 4 above)
Comment (pauldambra) [reply frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx:239]: ELI5 why you prefer that? being explicit?

(not challenging just for clarity 😊)

Comment (pauldambra) [reply frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx:239]: Screenshot 2024-06-10 at 09 45 16

the funny thing was `IFrameErrorOverlay` already internally managed its conditional display so could be simplified like this

Comment (daibhin) [reply frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx:239]: 1. Mainly consistency I guess. The neighbouring component is very similar but is conditional outside the component. I might find myself assuming that the `LoadingOverlay` is always rendered if I based it off how I see the `IframeErrorOverlay` works. 
2. Also makes the component stateless (won't always be true but is the case here). In theory should make it more reusable
3. It's also easier to reason about the size of the component if it is always rendered. Otherwise you have to remember as an implementor that the component can sometimes render nothing and account for that in your wrapping layout. Not going to miss that if the null case has to be accounted for in your implementation.
4. Testing is easier because you can pass values in

Comment (daibhin) [reply frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx:239]: Ah, I missed that 😄 I would say either works (but go with one approach in this component given the proximity) because I don't think we will achieve / should optimise for this kind of consistency across the codebase

Comment (pauldambra) [reply frontend/src/scenes/heatmaps/HeatmapsBrowser.tsx:239]: i'll let this seep into my subconscious... i've given up hoping that React will die so I guess I should develop some opinions 🤣
